File: src/aws/lambdas/incremental_distribution/src/main/java/uk/nhs/nhsx/virology/VirologyConfig.java
package uk.nhs.nhsx.virology;

import java.util.Objects;

public class VirologyConfig {

    public static final int MAX_TOKEN_PERSISTENCE_RETRY_COUNT = 3;

    public final String testOrdersTable;
    public final String testResultsTable;
    public final String submissionTokensTable;
    public final int maxTokenPersistenceRetryCount;
    public final String testOrdersIndex;

    public VirologyConfig(String testOrdersTable,
                          String testResultsTable,
                          String submissionTokensTable,
                          String testOrdersIndex,
                          int maxTokenPersistenceRetryCount) {
        if (maxTokenPersistenceRetryCount < 1) {
            throw new IllegalArgumentException("Retry count for persistence must be >= 1");
        }
        this.testOrdersTable = testOrdersTable;
        this.testResultsTable = testResultsTable;
        this.submissionTokensTable = submissionTokensTable;
        this.maxTokenPersistenceRetryCount = maxTokenPersistenceRetryCount;
        this.testOrdersIndex = testOrdersIndex;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        VirologyConfig that = (VirologyConfig) o;
        return maxTokenPersistenceRetryCount == that.maxTokenPersistenceRetryCount &&
            Objects.equals(testOrdersTable, that.testOrdersTable) &&
            Objects.equals(testResultsTable, that.testResultsTable) &&
            Objects.equals(submissionTokensTable, that.submissionTokensTable) &&
            Objects.equals(testOrdersIndex, that.testOrdersIndex);

    }

    @Override
    public int hashCode() {
        return Objects.hash(
            testOrdersTable,
            testResultsTable,
            submissionTokensTable,
            maxTokenPersistenceRetryCount,
            testOrdersIndex
        );
    }
}

File: src/aws/lambdas/incremental_distribution/src/main/java/uk/nhs/nhsx/virology/VirologyConfig.kt
package uk.nhs.nhsx.virology

import uk.nhs.nhsx.core.Environment
import uk.nhs.nhsx.core.Environment.EnvironmentKey
import uk.nhs.nhsx.core.EnvironmentKeys

data class VirologyConfig(val testOrdersTable: String,
                          val testResultsTable: String,
                          val submissionTokensTable: String,
                          val testOrdersIndex: String) {

    val maxTokenPersistenceRetryCount = 3

    companion object {

        private val TEST_ORDERS_TABLE = EnvironmentKey.string("test_orders_table")
        private val TEST_RESULTS_TABLE = EnvironmentKey.string("test_results_table")
        private val TEST_ORDERS_INDEX = EnvironmentKey.string("test_orders_index")

        fun fromEnvironment(environment: Environment): VirologyConfig {
            return VirologyConfig(
                environment.access.required(TEST_ORDERS_TABLE),
                environment.access.required(TEST_RESULTS_TABLE),
                environment.access.required(EnvironmentKeys.SUBMISSIONS_TOKENS_TABLE),
                environment.access.required(TEST_ORDERS_INDEX)
            )
        }
    }

}
